Automattic/wp-calypso
# A/B test implementation for Professional Email post-checkout upsell

#### Changes proposed in this Pull Request

* This PR implements the code for an A/B test of the post-checkout Professional Email upsell, where most of the work to render the upsell has already been implemented in previous PRs.
* The key change in this PR is to implement a new `/checkout/offer/:upsellType/:upsellMeta/:receiptId/:siteSlug` path and `PostCheckoutUpsellRedirector` component that uses `upsellType` to look up an experiment assignment and then redirect the user to an upsell or the Thank You page based on . For now, this code only handles the Professional Email upsell, but could easily be adapted to another upsell and/or a path that doesn't include `upsellMeta`.
  - It's worth noting that this particular change ensures that we're looking at a user who was both eligible for an upsell at the time they loaded checkout _and_ completed checkout, so we have no need for an exposure event.

The discussion includes some key input, primarily in the following comments:
* [when we ought to load information for checkout](https://github.com/Automattic/wp-calypso/pull/57195#issuecomment-947753421)
* [when we ought to be performing assignments](https://github.com/Automattic/wp-calypso/pull/57195#issuecomment-948202626)
* [when the post-checkout URL needs to be computed](https://github.com/Automattic/wp-calypso/pull/57195#issuecomment-948230036)
* [when we ought to assign relative to branching](https://github.com/Automattic/wp-calypso/pull/57195#issuecomment-950940118) (and also some limitations on exposure events)

Those comments in particular have informed the implementation I am using at the moment:
 * We're only sending users to the post checkout upsell redirector if they're eligible for the Professional Email offer
 * We're only performing the experiment assignment when we load the upsell redirector (which implies the user has checked out already)
 * The redirection is relatively quick, so should not heavily impact the UX for users
   - Right now, I'm not showing any particular UI in this space -- we can fix that if need be

#### Testing instructions

* Navigate to Experiment 20319 (only accessible to Automattic staff).
* Scroll down to the "Audience" section and mouse over the Bookmarklet link for either the `control` or `treatment` assignment.
* Click on the link in the popover and click on the link to assign yourself.
* Keep this page open in a tab/window - we're going to switch the assignment later.
* If you wish, you can enable Store Sandbox.
* Identify a site you own that has the free plan or a plan below the ECommerce plan. (You can cancel your plan if need be.)
* Identify a domain on that site that is eligible for the free Professional Email trial. You can also follow these steps to purchase a new domain that will be eligible:
  - Navigate to Upgrades -> Domains.
  - Click on the "Add a domain to this site" button, and then click on "Search for a domain" in the dropdown.
  - Feel free to perform one or more additional searches to find a test domain.
  - Click on "Select" next to a domain.
  - Click on "Skip" when you see the email upsell.
  - Complete checkout and purchase the domain
* Go to Upgrades -> Plans
* Select a paid plan other than the Premium plan to upgrade to -- it can be annual or monthly.
* Complete checkout using your preferred payment method.
  - _Ideally, we should make sure we test both a synchronous and a redirect-based payment method here, but I am not sure how best to test the asynchronous option._
* Verify that you're briefly redirected to `/checkout/offer/professional-email-offer/:domainName/:receiptId/:siteSlug`.
* If you are assigned to the `control` variation, verify that you are shown the standard plan Thank You page, except if you purchased the ECommerce plan, in which case you should see the start of the ECommerce onboarding flow.
* If you are assigned to the `treatment` variation, verify that you are shown the Professional Email upsell offer.

We need to repeat the tests above with the other variation (though you can use the same domain if you didn't complete the Professional Email purchase). Ideally, we also want to repeat the tests with another plan and/or payment method, as we want to cover the following cases:
 * Synchronous and redirect-based payment methods
 * Treatment and control variations of the experience
 * The ECommerce and non-ECommerce plans (as the ECommerce plan has a different post-checkout experience)

**NOTE:** Between each test/run, please ensure that you clear localStorage (mainly the `explat-experiment--promote_professional_email_post_checkout_2021_10` key) and the `calypso_store` under `IndexedDB` to ensure you get the updated experiment assignment.
